### PR TITLE
refactor(organizations): consolidation of organization controllers

### LINKED ISSUE #681: [FEAT] - Combine Organization controllers
### Priority

(Medium) I'm annoyed but I'll live

### Description

The `DexController`, `OrganizationController`, `RBACController`, `ServiceProxyController` and `TeamRoleSeederController` are reconciling the `Organization`.

Currently there is a workaround in place to ensure they do not interfere with each other when setting the `Ready` condition on the `Organization`.
With this PR the Controllers should be combined into a single controller, with a flow such as:

```mermaid
flowchart TD
  A[Organization Namepace]
  B@{ shape: diamond, label: "Created?" }
  C[AdminTeam]
  D[Organization RBAC]
  E[ServiceProxy]
  F[TeamRoles]
  G[Dex Connector]
  H@{ shape: diamond, label: "OIDC Config?" }
  J@{ shape: diamond, label: "ServiceProxy PluginDefinition?" }

  I[Return ERROR]
  K[Return]

  A --> B
  B -->|Yes| D
  B -->|No| I
  D --> F
  F --> J
  J -->|Yes| E
  E --> H
  J -->|No| H
  H -->|Yes| G
  G --> C
  H -->|No| K
  C --> K
```

Each of the Steps should set a `statusCondition` on the Organization to track the status of the Organization resource.

### Reference Issues

_No response_

### PR DESCRIPTION
**Pull request by gciezkowski-acc**
## Description
Consolidation of the organization controller.

## What type of PR is this? (check all applicable)

- [ ] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [X] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

- Related Issue # (issue)
- Closes #681 
- Fixes # (issue)

> Remove if not applicable

## Added tests?

- [ ] 👍 yes
- [X] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for tests # (issue/pr)

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Added to documentation?

- [ ] 📜 README.md
- [ ] 🤝 Documentation pages updated
- [X] 🙅 no documentation needed
- [ ] (if applicable) generated OpenAPI docs for CRD changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: pkg/controllers/organization/dex_controller.go [hunk 1]
```diff
@@ -51,52 +40,7 @@ type DexReconciler struct {
...
-	r.recorder = mgr.GetEventRecorderFor(name)
-	if r.Namespace == "" {
-		return errors.New("namespace required but missing")
-	}
-	return ctrl.NewControllerManagedBy(mgr).
-		Named(name).
-		For(&greenhousesapv1alpha1.Organization{},
-			builder.WithPredicates(clientutil.PredicateHasOICDConfigured())).
-		Owns(&dexapi.Connector{}).
-		Owns(&dexapi.OAuth2Client{}).
```
Comment (IvoGoman) [inline pkg/controllers/organization/dex_controller.go:66]: These need to be added to the OrganizationReconciler as well.

File: pkg/controllers/organization/rbac_controller.go [hunk 2]
```diff
@@ -31,92 +26,7 @@ const (
...
-func (r *RBACReconciler) SetupWithManager(name string, mgr ctrl.Manager) error {
-	r.Client = mgr.GetClient()
-	r.recorder = mgr.GetEventRecorderFor(name)
-	return ctrl.NewControllerManagedBy(mgr).
-		Named(name).
-		For(&greenhouseapisv1alpha1.Organization{}).
-		Owns(&rbacv1.Role{}).
-		Owns(&rbacv1.RoleBinding{}).
-		Owns(&rbacv1.ClusterRole{}).
-		Owns(&rbacv1.ClusterRoleBinding{}).
```
Comment (IvoGoman) [inline pkg/controllers/organization/rbac_controller.go:56]: These should move the the OrganizationReconciler's `SetupWithManager(..)`

File: pkg/controllers/organization/suite_test.go [hunk 3]
```diff
@@ -30,9 +30,9 @@ var _ = BeforeSuite(func() {
 	groupsServer = scim.ReturnDefaultGroupResponseMockServer()
 
 	test.RegisterController("organizationController", (&organizationpkg.OrganizationReconciler{}).SetupWithManager)
-	test.RegisterController("organizationRBAC", (&organizationpkg.RBACReconciler{}).SetupWithManager)
-	test.RegisterController("organizationServiceProxy", (&organizationpkg.ServiceProxyReconciler{}).SetupWithManager)
-	test.RegisterController("teamRoleSeeder", (&organizationpkg.TeamRoleSeederReconciler{}).SetupWithManager)
+	// test.RegisterController("organizationRBAC", (&organizationpkg.RBACReconciler{}).SetupWithManager)
+	// test.RegisterController("organizationServiceProxy", (&organizationpkg.ServiceProxyReconciler{}).SetupWithManager)
+	// test.RegisterController("teamRoleSeeder", (&organizationpkg.TeamRoleSeederReconciler{}).SetupWithManager)
```
Comment (IvoGoman) [inline pkg/controllers/organization/suite_test.go:35]: These lines can be removed

**Review by IvoGoman — CHANGES_REQUESTED** [review]
Looks good already. 
Couple of things we can improve while refactoring the OrganizationController.

1. StatusConditions for the different Steps (e.g. `OrganizationRBACConfigured`, `ServiceProxyProvisioned`, `TeamRolesSeeded`, `OrganizationOIDCCnfigured`) This way we know what the current state of the Organization is
2. `git mv dex_controller.go dex.go` to better distinguish between the controller and

File: pkg/controllers/organization/organization_controller.go [hunk 4]
```diff
@@ -31,15 +34,23 @@ var (
...
 //+kubebuilder:rbac:groups=greenhouse.sap,resources=organizations/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=greenhouse.sap,resources=organizations/finalizers,verbs=update
 //+kubebuilder:rbac:groups=greenhouse.sap,resources=teams,verbs=get;watch;create;update;patch
+//+kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=clusterroles;clusterrolebindings;roles;rolebindings,verbs=get;list;watch;create;update;patch
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=plugindefinitions,verbs=get;list;watch
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=plugins,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=teamroles,verbs=get;list;watch;create;update;patch
 //+kubebuilder:rbac:groups="",resources=events,verbs=get;list;watch;create;update;patch
 //+kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch;create;update;patch
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=organizations,verbs=get;list;watch
```
[the remark this answers is not in the document]
Comment (gciezkowski-acc) [reply pkg/controllers/organization/organization_controller.go]: Done

File: pkg/controllers/organization/organization_controller.go [hunk 5]
```diff
@@ -77,6 +99,32 @@ func (r *OrganizationReconciler) EnsureCreated(ctx context.Context, object lifec
...
+		}
+
+		if err := r.reconcileOAuth2Client(ctx, org); err != nil {
+			return ctrl.Result{}, lifecycle.Failed, err
+		}
+	}
+
+	if err := r.reconcileAdminTeam(ctx, org); err != nil {
+		return ctrl.Result{}, lifecycle.Failed, err
+	}
```
[the remark this answers is not in the document]
Comment (gciezkowski-acc) [reply pkg/controllers/organization/organization_controller.go]: Done

File: pkg/controllers/organization/dex_controller.go [hunk 6]
```diff
@@ -18,31 +18,20 @@ import (
...
 const dexConnectorTypeGreenhouse = "greenhouse-oidc"
 
-// DexReconciler reconciles a Organization object
-type DexReconciler struct {
-	client.Client
-	recorder  record.EventRecorder
-	Namespace string
-}
-
 //+kubebuilder:rbac:groups=greenhouse.sap,resources=organizations,verbs=get;list;watch;create;update;patch;delete
```
[the remark this answers is not in the document]
Comment (gciezkowski-acc) [reply pkg/controllers/organization/dex_controller.go:35]: Done

File: pkg/controllers/organization/rbac_controller.go (hunk 2 above)
Comment (gciezkowski-acc) [reply pkg/controllers/organization/rbac_controller.go:56]: Done

File: pkg/controllers/organization/suite_test.go (hunk 3 above)
Comment (gciezkowski-acc) [reply pkg/controllers/organization/suite_test.go]: Done

File: pkg/controllers/organization/dex_controller.go (hunk 1 above)
Comment (gciezkowski-acc) [reply pkg/controllers/organization/dex_controller.go:66]: Done

**Comment by gciezkowski-acc** [comment]
> @gciezkowski-acc drop the `_controller` for the other files, as they are not controllers anymore. They are just segregated functionalities.

Done

**Review by abhijith-darshan — DISMISSED** [review]
Looks good from my side. Great work!

if this feedback is incorporated then we can merge it. 
https://github.com/cloudoperators/greenhouse/pull/744#pullrequestreview-2445624227

File: pkg/apis/greenhouse/v1alpha1/organization_types.go [hunk 7]
```diff
@@ -16,6 +16,23 @@ const (
...
+	OrganizationRBACConfigured ConditionType = "OrganizationRBACConfigured"
+	// OrganizationDefaultTeamRoleConfigured is set when default team roles are configured
+	OrganizationDefaultTeamRoleConfigured ConditionType = "OrganizationDefaultTeamRoleConfigured"
+	// ServiceProxyProvisioned is set when the service proxy is provisioned
+	ServiceProxyProvisioned ConditionType = "ServiceProxyProvisioned"
+	// OrganizationOICDConfigured is set when the OICD is configured
+	OrganizationOICDConfigured ConditionType = "OrganizationOICDConfigured"
+	// DexReconcileFailed is set when dex reconcile step failed
+	DexReconcileFailed ConditionReason = "DexReconcileFailed"
+	// OAuthOICDFailed is set when OAuth reconciler failed
```
Comment (abhijith-darshan) [inline pkg/apis/greenhouse/v1alpha1/organization_types.go:32]:
```suggestion
	// OAuthOICDFailed is set when OAuth reconcile step has failed
```

File: pkg/apis/greenhouse/v1alpha1/organization_types.go [hunk 8]
```diff
@@ -16,6 +16,23 @@ const (
...
+	NamespacePreparedCondition ConditionType = "NamespacePrepared"
+	// OrganizationRBACConfigured is set when the RBAC for organization is configured
+	OrganizationRBACConfigured ConditionType = "OrganizationRBACConfigured"
+	// OrganizationDefaultTeamRoleConfigured is set when default team roles are configured
+	OrganizationDefaultTeamRoleConfigured ConditionType = "OrganizationDefaultTeamRoleConfigured"
+	// ServiceProxyProvisioned is set when the service proxy is provisioned
+	ServiceProxyProvisioned ConditionType = "ServiceProxyProvisioned"
+	// OrganizationOICDConfigured is set when the OICD is configured
+	OrganizationOICDConfigured ConditionType = "OrganizationOICDConfigured"
+	// DexReconcileFailed is set when dex reconcile step failed
```
Comment (abhijith-darshan) [inline pkg/apis/greenhouse/v1alpha1/organization_types.go:30]:
```suggestion
	// DexReconcileFailed is set when dex reconcile step has failed
```

**Review by abhijith-darshan — CHANGES_REQUESTED** [review]
some NITs

File: pkg/apis/greenhouse/v1alpha1/organization_types.go [hunk 9]
```diff
@@ -16,6 +16,23 @@ const (
 	SCIMRequestFailedReason ConditionReason = "SCIMRequestFailed"
 	// SCIMConfigNotProvidedReason is set when scim config is not present in spec as it is optional
 	SCIMConfigNotProvidedReason ConditionReason = "SCIMConfigNotProvided"
+
+	// NamespacePreparedCondition is set when the namespace for organization is prepared.
```
[the remark this answers is not in the document]
Comment (gciezkowski-acc) [reply pkg/apis/greenhouse/v1alpha1/organization_types.go]: Done

File: pkg/apis/greenhouse/v1alpha1/organization_types.go (hunk 8 above)
Comment (gciezkowski-acc) [reply pkg/apis/greenhouse/v1alpha1/organization_types.go]: Done

File: pkg/apis/greenhouse/v1alpha1/organization_types.go [hunk 10]
```diff
@@ -16,6 +16,23 @@ const (
 	SCIMRequestFailedReason ConditionReason = "SCIMRequestFailed"
 	// SCIMConfigNotProvidedReason is set when scim config is not present in spec as it is optional
 	SCIMConfigNotProvidedReason ConditionReason = "SCIMConfigNotProvided"
+
+	// NamespaceCreated is set when the namespace for organization is created.
+	NamespaceCreated ConditionType = "NamespaceCreated"
+	// OrganizationRBACConfigured is set when the RBAC for organization is configured
+	OrganizationRBACConfigured ConditionType = "OrganizationRBACConfigured"
+	// OrganizationDefaultTeamRoleConfigured is set when default team roles are configured
+	OrganizationDefaultTeamRoleConfigured ConditionType = "OrganizationDefaultTeamRoleConfigured"
```
Comment (IvoGoman) [inline pkg/apis/greenhouse/v1alpha1/organization_types.go:25]:
```suggestion
	// OrganizationDefaultTeamRolesConfigured is set when default team roles are configured
	OrganizationDefaultTeamRolesConfigured ConditionType = "OrganizationDefaultTeamRolesConfigured"
```

**Review by IvoGoman — APPROVED** [review]

**Review by abhijith-darshan — APPROVED** [review]

**Pull request merged**